### DISCUSSION TITLE
PSA regarding ASan and ASLR on Linux

### DISCUSSION THREAD

**Discussion by JohnoKing** [ksh93/ksh#702] [General]
I recently noticed that attempting to compile ksh with ASan enabled now fails on Arch Linux with the following error:
```
AddressSanitizer:DEADLYSIGNAL
=================================================================
==29986==ERROR: AddressSanitizer: SEGV on unknown address 0x61c626338e78 (pc 0x75088e03338f bp 0x000000000000 sp 0x7ffc34150ff0 T0)
==29986==The signal is caused by a READ memory access.
AddressSanitizer:DEADLYSIGNAL
AddressSanitizer: nested bug in the same thread, aborting.
mamake: *** exit code 1 making lib/libast
mamake: *** exit code 1 making all
package: make failed at Sun Jan 14 10:42:40 PST 2024 in /home/johno/GitRepos/KornShell/ksh/arch/linux.i386-64
```

This failure is the result of a mitigation for https://zolutal.github.io/aslrnt/ (see https://gitlab.archlinux.org/archlinux/packaging/packages/linux-lts/-/commit/de42b5567). I'm not filing this as a bug report because I think the crash is the result of ASan being inherently incompatible with  ASLR (i.e., probably not a bug).
In any case, compiling ksh with ASan now requires that ASLR be disabled. On Linux this can be done as follows:
```sh
# Using 'setarch -R'
setarch $(uname -m) -R bin/package make CCFLAGS='-O0 -g -fsanitize=address'
ASAN_OPTIONS='detect_leaks=0' setarch $(uname -m) -R bin/shtests

# Using randomize_va_space
echo 0 | sudo tee /proc/sys/kernel/randomize_va_space
```
After you're done using ASan for testing, ASLR can be reenabled by setting randomize_va_space to '2'.

**Comment by McDutchie** [comment]
Thanks, I appreciate the heads-up.
